[Bug]: tests of PlayQueueService are unstable

### What happened?

Tests of PlayQueueService are unsable and CI for testing fails randomly.

### Steps to reproduce

1. Create PR and executes CI for testing.

### Version

11.1.x (Edge)

### Version Detail

_No response_

### Operating System

ubuntu-latest(GitHub Action)

### Java Version

17.0.5

### Database

PostgreSQL 13.x, PostgreSQL 14.x, MariaDB 8.x, Mysql 10.x

### DB Detail

_No response_

### Configuration paramter

Parameters are written here.

https://github.com/kagemomiji/airsonic-advanced/blob/5ae84133dddc0cd111cefe7b0a0831fd49c42d54/.github/workflows/pr_ci.yml#L44-L122

### Proxy Server

_No response_

### client detail

_No response_

### language

None

### Relevant log output

```console
[INFO] 
[INFO] Results:
[INFO] 
Error:  Errors: 
Error:  org.airsonic.player.service.PlayQueueServiceTest$NonJukeboxTests.testStop
[INFO]   Run 1: PASS
Error:    Run 2: PlayQueueServiceTest$NonJukeboxTests.testStop:148 » Execution Wanted but not i...
[INFO] 
[INFO] 
Error:  Tests run: 425, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Airsonic 11.1.3-SNAPSHOT:
[INFO] 
[INFO] Airsonic ........................................... SUCCESS [  1.086 s]
[INFO] Subsonic REST API .................................. SUCCESS [  4.144 s]
[INFO] Sonos API .......................................... SUCCESS [  4.332 s]
[INFO] Airsonic Main ...................................... FAILURE [09:00 min]
[INFO] Airsonic Docker Image .............................. SKIPPED
[INFO] Airsonic Integration Test .......................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  09:10 min
[INFO] Finished at: 2023-07-01T12:35:00Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project airsonic-main: There are test failures.
Error:  
Error:  Please refer to /home/runner/work/airsonic-advanced/airsonic-advanced/airsonic-main/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :airsonic-main
Error: Process completed with exit code 1.
```

--- a/airsonic-main/src/main/java/org/airsonic/player/config/ThreadPoolConfig.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/config/ThreadPoolConfig.java
@@ -1,0 +1,23 @@
+package org.airsonic.player.config;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
+
+import java.util.concurrent.Executor;
+
+@Configuration
+public class ThreadPoolConfig {
+
+    @Bean(name = "BroadcastThreadPool")
+    public Executor configThreadPool() {
+        var threadPool = new ThreadPoolTaskExecutor();
+        threadPool.setCorePoolSize(2);
+        threadPool.setMaxPoolSize(5);
+        threadPool.setQueueCapacity(500);
+        threadPool.setDaemon(true);
+        threadPool.setThreadNamePrefix("BroadcastThread-");
+        threadPool.initialize();
+        return threadPool;
+    }
+}

--- a/airsonic-main/src/main/java/org/airsonic/player/service/PlayQueueService.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/service/PlayQueueService.java
@@ -17,10 +17,10 @@ import org.airsonic.player.domain.PodcastEpisode;
 import org.airsonic.player.domain.PodcastStatus;
 import org.airsonic.player.domain.RandomSearchCriteria;
 import org.airsonic.player.domain.SavedPlayQueue;
+import org.airsonic.player.service.websocket.AsyncWebSocketClient;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.messaging.simp.SimpMessageHeaderAccessor;
 import org.springframework.messaging.simp.SimpMessageType;
-import org.springframework.messaging.simp.SimpMessagingTemplate;
 import org.springframework.stereotype.Service;
 
 import java.time.Instant;
@@ -29,9 +29,8 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Function;
-
-import static java.util.concurrent.CompletableFuture.runAsync;
 
 @Service
 public class PlayQueueService {
@@ -64,15 +63,15 @@ public class PlayQueueService {
     @Autowired
     private InternetRadioService internetRadioService;
     @Autowired
-    private SimpMessagingTemplate brokerTemplate;
+    private AsyncWebSocketClient webSocketClient;
 
     public void start(Player player) {
         player.getPlayQueue().setStatus(PlayQueue.Status.PLAYING);
         if (player.isJukebox()) {
             jukeboxService.start(player);
         }
-        runAsync(() -> brokerTemplate.convertAndSendToUser(player.getUsername(),
-                "/queue/playqueues/" + player.getId() + "/playstatus", PlayQueue.Status.PLAYING));
+        webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/playstatus",
+                PlayQueue.Status.PLAYING);
     }
 
     public void stop(Player player) {
@@ -80,8 +79,8 @@ public class PlayQueueService {
         if (player.isJukebox()) {
             jukeboxService.stop(player);
         }
-        runAsync(() -> brokerTemplate.convertAndSendToUser(player.getUsername(),
-                "/queue/playqueues/" + player.getId() + "/playstatus", PlayQueue.Status.STOPPED));
+        webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/playstatus",
+                PlayQueue.Status.STOPPED);
     }
 
     public void toggleStartStop(Player player) {
@@ -98,12 +97,11 @@ public class PlayQueueService {
             jukeboxService.skip(player, index, (int) (offset / 1000));
         }
 
-        runAsync(() -> {
-            brokerTemplate.convertAndSendToUser(player.getUsername(),
-                    "/queue/playqueues/" + player.getId() + "/skip", ImmutableMap.of("index", index, "offset", offset));
-            brokerTemplate.convertAndSendToUser(player.getUsername(),
-                    "/queue/playqueues/" + player.getId() + "/playstatus", player.getPlayQueue().getStatus());
-        });
+        CompletableFuture<Void> resultSkip = webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/skip",
+                ImmutableMap.of("index", index, "offset", offset));
+
+        resultSkip.thenRun(() -> webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/playstatus",
+                player.getPlayQueue().getStatus()));
     }
 
     public void reloadSearchCriteria(Player player, String sessionId) {
@@ -437,8 +435,8 @@ public class PlayQueueService {
         } else {
             playQueue.setRepeatStatus(RepeatStatus.getNext(playQueue.getRepeatStatus()));
         }
-        runAsync(() -> brokerTemplate.convertAndSendToUser(player.getUsername(),
-                "/queue/playqueues/" + player.getId() + "/repeat", playQueue.getRepeatStatus()));
+        webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/repeat",
+                playQueue.getRepeatStatus());
     }
 
     public void undo(Player player) {
@@ -459,14 +457,12 @@ public class PlayQueueService {
     //
     public void setJukeboxGain(Player player, float gain) {
         jukeboxService.setGain(player, gain);
-        runAsync(() -> brokerTemplate.convertAndSendToUser(player.getUsername(),
-                "/queue/playqueues/" + player.getId() + "/jukebox/gain", gain));
+        webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/jukebox/gain", gain);
     }
 
     public void setJukeboxPosition(Player player, int positionInSeconds) {
         jukeboxService.setPosition(player, positionInSeconds);
-        runAsync(() -> brokerTemplate.convertAndSendToUser(player.getUsername(),
-                "/queue/playqueues/" + player.getId() + "/jukebox/position", positionInSeconds));
+        webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/jukebox/position", positionInSeconds);
     }
 
     //
@@ -478,12 +474,9 @@ public class PlayQueueService {
     }
 
     private void broadcastPlayQueue(Player player, Function<PlayQueueInfo, PlayQueueInfo> playQueueModifier, String triggeringSessionId) {
-        runAsync(() -> {
-            PlayQueueInfo info = playQueueModifier.apply(getPlayQueueInfo(player, ""));
-            brokerTemplate.convertAndSendToUser(player.getUsername(),
-                    "/queue/playqueues/" + player.getId() + "/updated", info);
-            postBroadcast(info, player, triggeringSessionId);
-        });
+        PlayQueueInfo info = playQueueModifier.apply(getPlayQueueInfo(player, ""));
+        CompletableFuture<Void> updatedFuture = webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/updated", info);
+        updatedFuture.thenRun(() -> postBroadcast(info, player, triggeringSessionId));
     }
 
     private void postBroadcast(PlayQueueInfo info, Player player, String sessionId) {
@@ -492,8 +485,7 @@ public class PlayQueueService {
                 // trigger the web player to start playing at this location
                 SimpMessageHeaderAccessor headerAccessor = SimpMessageHeaderAccessor.create(SimpMessageType.MESSAGE);
                 headerAccessor.setSessionId(sessionId);
-                brokerTemplate.convertAndSendToUser(player.getUsername(),
-                        "/queue/playqueues/" + player.getId() + "/skip",
+                webSocketClient.sendToUser(player.getUsername(), "/queue/playqueues/" + player.getId() + "/skip",
                         ImmutableMap.of("index", info.getStartPlayerAt(), "offset", info.getStartPlayerAtPosition()),
                         headerAccessor.getMessageHeaders());
             } else if (!player.isExternalWithPlaylist()) {

--- a/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java
@@ -34,7 +34,7 @@ import org.airsonic.player.domain.PodcastStatus;
 import org.airsonic.player.service.metadata.MetaData;
 import org.airsonic.player.service.metadata.MetaDataParser;
 import org.airsonic.player.service.metadata.MetaDataParserFactory;
-import org.airsonic.player.service.websocket.AsyncSocketClient;
+import org.airsonic.player.service.websocket.AsyncWebSocketClient;
 import org.airsonic.player.util.FileUtil;
 import org.airsonic.player.util.StringUtil;
 import org.airsonic.player.util.Util;
@@ -109,7 +109,7 @@ public class PodcastService {
     private final MetaDataParserFactory metaDataParserFactory;
     private final VersionService versionService;
     private final TaskSchedulingService taskService;
-    private final AsyncSocketClient asyncSocketClient;
+    private final AsyncWebSocketClient asyncSocketClient;
     private Runnable defaultTask;
     private Predicate<PodcastEpisode> filterAllowed;
 
@@ -123,7 +123,7 @@ public class PodcastService {
         MetaDataParserFactory metaDataParserFactory,
         VersionService versionService,
         TaskSchedulingService taskService,
-        AsyncSocketClient asyncSocketClient
+        AsyncWebSocketClient asyncSocketClient
     ) {
         this.podcastDao = podcastDao;
         this.settingsService = settingsService;

--- a/airsonic-main/src/main/java/org/airsonic/player/service/websocket/AsyncWebSocketClient.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/service/websocket/AsyncWebSocketClient.java
@@ -29,9 +29,9 @@ import org.springframework.stereotype.Component;
 
 @Component
 @EnableAsync
-public class AsyncSocketClient {
+public class AsyncWebSocketClient {
 
-    private final Logger LOG = LoggerFactory.getLogger(AsyncSocketClient.class);
+    private final Logger LOG = LoggerFactory.getLogger(AsyncWebSocketClient.class);
 
     @Autowired
     private SimpMessagingTemplate brokerTemplate;

--- a/airsonic-main/src/main/java/org/airsonic/player/service/websocket/AsyncWebSocketClient.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/service/websocket/AsyncWebSocketClient.java
@@ -22,13 +22,17 @@ package org.airsonic.player.service.websocket;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.AdviceMode;
 import org.springframework.messaging.simp.SimpMessagingTemplate;
 import org.springframework.scheduling.annotation.Async;
 import org.springframework.scheduling.annotation.EnableAsync;
 import org.springframework.stereotype.Component;
 
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
 @Component
-@EnableAsync
+@EnableAsync(mode = AdviceMode.ASPECTJ)
 public class AsyncWebSocketClient {
 
     private final Logger LOG = LoggerFactory.getLogger(AsyncWebSocketClient.class);
@@ -36,9 +40,23 @@ public class AsyncWebSocketClient {
     @Autowired
     private SimpMessagingTemplate brokerTemplate;
 
-    @Async
+    @Async("BroadcastThreadPool")
     public void send(String destination, Object payload) {
         LOG.debug("Sending to {}: {}", destination, payload);
         brokerTemplate.convertAndSend(destination, payload);
+    }
+
+    @Async("BroadcastThreadPool")
+    public CompletableFuture<Void> sendToUser(String user, String destination, Object payload) {
+        LOG.debug("Sending to user {}: {}: {}", user, destination, payload);
+        brokerTemplate.convertAndSendToUser(user, destination, payload);
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Async("BroadcastThreadPool")
+    public CompletableFuture<Void> sendToUser(String user, String destination, Object payload, Map<String, Object> headers) {
+        LOG.debug("Sending to user {}: {}: {}", user, destination, payload);
+        brokerTemplate.convertAndSendToUser(user, destination, payload, headers);
+        return CompletableFuture.completedFuture(null);
     }
 }

--- a/airsonic-main/src/test/java/org/airsonic/player/service/PlayQueueServiceTest.java
+++ b/airsonic-main/src/test/java/org/airsonic/player/service/PlayQueueServiceTest.java
@@ -24,6 +24,7 @@ import org.airsonic.player.domain.MediaFile;
 import org.airsonic.player.domain.PlayQueue;
 import org.airsonic.player.domain.Player;
 import org.airsonic.player.domain.SavedPlayQueue;
+import org.airsonic.player.service.websocket.AsyncWebSocketClient;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Nested;
@@ -36,13 +37,11 @@ import org.mockito.ArgumentCaptor;
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
-import org.springframework.messaging.simp.SimpMessagingTemplate;
 
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ScheduledThreadPoolExecutor;
-import java.util.concurrent.TimeUnit;
 
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
@@ -52,7 +51,6 @@ import static org.mockito.ArgumentMatchers.anyInt;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.timeout;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
@@ -62,7 +60,7 @@ public class PlayQueueServiceTest {
     @Mock
     private JukeboxService jukeboxService;
     @Mock
-    private SimpMessagingTemplate brokerTemplate;
+    private AsyncWebSocketClient webSocketClient;
     @Mock
     private MediaFileService mediaFileService;
     @Mock
@@ -105,22 +103,15 @@ public class PlayQueueServiceTest {
             when(mockedPlayer.getUsername()).thenReturn("testuser");
             when(mockedPlayer.getId()).thenReturn(1);
 
-            // Verify brokerTemplate is called to send message
-            // Configure mock behavior
-            doAnswer(invocation -> {
-                TimeUnit.MILLISECONDS.sleep(100);
-                return null;
-            }).when(brokerTemplate).convertAndSendToUser(anyString(), anyString(), any());
+            // Verify websocketClient is called to send message
+            when(webSocketClient.sendToUser(anyString(), anyString(), any())).thenReturn(CompletableFuture.completedFuture(null));
 
             // Test
             playQueueService.start(mockedPlayer);
 
             // then
             verify(mockedPlayQueue).setStatus(PlayQueue.Status.PLAYING);
-            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
-                verify(brokerTemplate).convertAndSendToUser("testuser", "/queue/playqueues/1/playstatus", PlayQueue.Status.PLAYING);
-            }, executor);
-            future.get(3000, TimeUnit.MILLISECONDS); // Wait for the future to complete
+            verify(webSocketClient).sendToUser("testuser", "/queue/playqueues/1/playstatus", PlayQueue.Status.PLAYING);
         }
 
         @Test
@@ -130,22 +121,15 @@ public class PlayQueueServiceTest {
             when(mockedPlayer.getUsername()).thenReturn("testuser");
             when(mockedPlayer.getId()).thenReturn(1);
 
-            // Verify brokerTemplate is called to send message
-            // Configure mock behavior
-            doAnswer(invocation -> {
-                TimeUnit.MILLISECONDS.sleep(100);
-                return null;
-            }).when(brokerTemplate).convertAndSendToUser(anyString(), anyString(), any());
+            // Verify websocketClient is called to send message
+            when(webSocketClient.sendToUser(anyString(), anyString(), any())).thenReturn(CompletableFuture.completedFuture(null));
 
             // Test
             playQueueService.stop(mockedPlayer);
 
             // then
             verify(mockedPlayQueue).setStatus(PlayQueue.Status.STOPPED);
-            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
-                verify(brokerTemplate).convertAndSendToUser("testuser", "/queue/playqueues/1/playstatus", PlayQueue.Status.STOPPED);
-            }, executor);
-            future.get(3000, TimeUnit.MILLISECONDS); // Wait for the future to complete
+            verify(webSocketClient).sendToUser("testuser", "/queue/playqueues/1/playstatus", PlayQueue.Status.STOPPED);
         }
 
         @ParameterizedTest
@@ -160,23 +144,17 @@ public class PlayQueueServiceTest {
             when(mockedPlayer.getId()).thenReturn(1);
             when(mockedPlayQueue.getStatus()).thenReturn(initialStatus);
 
-            // Verify brokerTemplate is called to send message
-            // Configure mock behavior
-            doAnswer(invocation -> {
-                TimeUnit.MILLISECONDS.sleep(100);
-                return null;
-            }).when(brokerTemplate).convertAndSendToUser(anyString(), anyString(), any());
+            // Verify websocketClient is called to send message
+            when(webSocketClient.sendToUser(anyString(), anyString(), any())).thenReturn(CompletableFuture.completedFuture(null));
 
             // Test
             playQueueService.toggleStartStop(mockedPlayer);
 
             // then
             verify(mockedPlayQueue).setStatus(expectedStatus);
-            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
-                verify(brokerTemplate).convertAndSendToUser("testuser", "/queue/playqueues/1/playstatus", expectedStatus);
-            }, executor);
-            future.get(3000, TimeUnit.MILLISECONDS); // Wait for the future to complete
+            verify(webSocketClient).sendToUser("testuser", "/queue/playqueues/1/playstatus", expectedStatus);
         }
+
         @Test
         public void testSkip() throws Exception {
             // given
@@ -184,14 +162,14 @@ public class PlayQueueServiceTest {
             when(mockedPlayer.getUsername()).thenReturn("testuser");
             when(mockedPlayer.getId()).thenReturn(1);
             when(mockedPlayQueue.getStatus()).thenReturn(PlayQueue.Status.PLAYING);
+            when(webSocketClient.sendToUser(anyString(), anyString(), any())).thenReturn(CompletableFuture.completedFuture(null));
 
             // when
-            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> playQueueService.skip(mockedPlayer, 2, 3L), executor);
-            future.get();
+            playQueueService.skip(mockedPlayer, 2, 3L);
 
             // then
-            verify(brokerTemplate, timeout(1000)).convertAndSendToUser("testuser", "/queue/playqueues/1/skip", ImmutableMap.of("index", 2, "offset", 3L));
-            verify(brokerTemplate, timeout(1000)).convertAndSendToUser("testuser", "/queue/playqueues/1/playstatus", PlayQueue.Status.PLAYING);
+            verify(webSocketClient).sendToUser("testuser", "/queue/playqueues/1/skip", ImmutableMap.of("index", 2, "offset", 3L));
+            verify(webSocketClient).sendToUser("testuser", "/queue/playqueues/1/playstatus", PlayQueue.Status.PLAYING);
             verify(mockedPlayQueue).setIndex(2);
         }
     }

--- a/airsonic-main/src/test/java/org/airsonic/player/service/PodcastServiceTestCase.java
+++ b/airsonic-main/src/test/java/org/airsonic/player/service/PodcastServiceTestCase.java
@@ -27,7 +27,7 @@ import org.airsonic.player.domain.PodcastChannelRule;
 import org.airsonic.player.domain.PodcastEpisode;
 import org.airsonic.player.domain.PodcastStatus;
 import org.airsonic.player.domain.Version;
-import org.airsonic.player.service.websocket.AsyncSocketClient;
+import org.airsonic.player.service.websocket.AsyncWebSocketClient;
 import org.apache.http.HttpEntity;
 import org.apache.http.client.methods.CloseableHttpResponse;
 import org.apache.http.impl.client.CloseableHttpClient;
@@ -85,7 +85,7 @@ public class PodcastServiceTestCase {
     @Mock
     private MediaFolderService mediaFolderService;
     @Mock
-    private AsyncSocketClient asyncSocketClient;
+    private AsyncWebSocketClient asyncSocketClient;
     @Mock
     private VersionService versionService;
 

